src/hotspot/share/opto/intrinsicnode.cpp

```diff
@@ -112,3 +114,109 @@ SignumFNode* SignumFNode::make(PhaseGVN& gvn, Node* in) {
   return new SignumFNode(in, gvn.makecon(TypeF::ZERO), gvn.makecon(TypeF::ONE));
 }
 
+Node* CompressBitsNode::Ideal(PhaseGVN* phase, bool can_reshape) {
+  Node* src = in(1);
+  Node* mask = in(2);
+  if (bottom_type()->isa_int()) {
+    if (mask->Opcode() == Op_LShiftI && phase->type(mask->in(1))->is_int()->is_con()) {
+      // compress(x, 1 << n) == (x >> n & 1)
+      if (phase->type(mask->in(1))->higher_equal(TypeInt::ONE)) {
+        Node* rshift = phase->transform(new RShiftINode(in(1), mask->in(2)));
+        return new AndINode(rshift, phase->makecon(TypeInt::ONE));
+      // compress(x, -1 << n) == x >>> n
+      } else if (phase->type(mask->in(1))->higher_equal(TypeInt::MINUS_1)) {
+        return new URShiftINode(in(1), mask->in(2));
+      }
+    }
+    // compress(expand(x, m), m) == x & compress(m, m)
+    if (src->Opcode() == Op_ExpandBits &&
+        src->in(2) == mask) {
+      Node* compr = phase->transform(new CompressBitsNode(mask, mask, TypeInt::INT));
+      return new AndINode(compr, src->in(1));
+    }
+  } else {
+    assert(bottom_type()->isa_long(), "");
+    if (mask->Opcode() == Op_LShiftL && phase->type(mask->in(1))->is_long()->is_con()) {
+      // compress(x, 1 << n) == (x >> n & 1)
+      if (phase->type(mask->in(1))->higher_equal(TypeLong::ONE)) {
+        Node* rshift = phase->transform(new RShiftLNode(in(1), mask->in(2)));
+        return new AndLNode(rshift, phase->makecon(TypeLong::ONE));
+      // compress(x, -1 << n) == x >>> n
+      } else if (phase->type(mask->in(1))->higher_equal(TypeLong::MINUS_1)) {
+        return new URShiftLNode(in(1), mask->in(2));
+      }
+    }
+    // compress(expand(x, m), m) == x & compress(m, m)
+    if (src->Opcode() == Op_ExpandBits &&
+        src->in(2) == mask) {
+      Node* compr = phase->transform(new CompressBitsNode(mask, mask, TypeLong::LONG));
+      return new AndLNode(compr, src->in(1));
+    }
+  }
+  return NULL;
+}
+
+Node* compress_expand_identity(PhaseGVN* phase, Node* n) {
+  if (n->bottom_type()->isa_int()) {
+    // compress(x, 0) == 0
+    if(phase->type(n->in(2))->higher_equal(TypeInt::ZERO)) return n->in(2);
+    // compress(x, -1) == x
+    if(phase->type(n->in(2))->higher_equal( TypeInt::MINUS_1)) return n->in(1);
+  } else {
+    assert(n->bottom_type()->isa_long(), "");
+    // compress(x, 0) == 0
+    if(phase->type(n->in(2))->higher_equal(TypeLong::ZERO)) return n->in(2);
+    // compress(x, -1) == x
+    if(phase->type(n->in(2))->higher_equal( TypeLong::MINUS_1)) return n->in(1);
+  }
```

Comment:
The codes are almost the same for int and long type except for the "ZERO and MINUS_1" node. Could you please remove the duplicate codes by just defining different `ZERO` and `MUNUS_1` nodes for int and long?

Reply:
One more style issue: `higher_equal( TypeLong::MINUS_1)  -> higher_equal(TypeLong::MINUS_1)`

Reply:
TypeInteger can work. https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/type.cpp#L1416